pytorch_lightning/trainer/distrib_data_parallel.py

```diff
@@ -379,28 +387,30 @@ def spawn_ddp_children(self, model):
         os.environ['NODE_RANK'] = node_rank
         os.environ['LOCAL_RANK'] = '0'
 
+        # when user is using hydra find the absolute path
+        path_lib = abspath if not HYDRA_AVAILABLE else to_absolute_path
+
         # pull out the commands used to run the script and resolve the abs file path
         command = sys.argv
-        full_path = abspath(command[0])
+        try:
+            full_path = path_lib(command[0])
+        except Exception as e:
+            full_path = abspath(command[0])
+
         command[0] = full_path
         command = ['python'] + command
 
         # since this script sets the visible devices we replace the gpus flag with a number
         num_gpus = os.environ['CUDA_VISIBLE_DEVICES'].split(',').__len__()
 
-        # if script called without a flag, pass in a flag anyhow
-        if '--gpus' not in command:
-            arg_gpus = len(self.gpus) if isinstance(self.gpus, list) else self.gpus
-            command += ['--gpus', arg_gpus]
-
-        gpu_flag_idx = command.index('--gpus')
-        command[gpu_flag_idx + 1] = f'{num_gpus}'
+        if '--gpus' in command:
+            gpu_flag_idx = command.index('--gpus')
+            command[gpu_flag_idx + 1] = f'{num_gpus}'
```

Comment:
I asked a few times already and did not get an answer:
Who is defining --gpus and who is consuming it?

If it's safe to drop for a Hydra app, why is it not safe to drop for any app and just rely on the environment variables always?

Reply:
--gpus is something that is defined by a pytorch lightning program and it is being consumer by a pytorch lightning program when you setup DDP. You can see the current structure used here: https://pytorch-lightning.readthedocs.io/en/latest/introduction_guide.html#argparser-best-practices. This comes back to using  these functions: https://github.com/PyTorchLightning/pytorch-lightning/blob/04c794ca72f45dd492e49baff773c2b8e274b9ed/pytorch_lightning/trainer/trainer.py#L616.

Reply:
Gotcha.
Since argparse best practices are not relevant when a user is using Hydra we need to figure out what is the best way to configure PL with Hydra.

@anthonytec2, would you be able to create a standalone example we can iterate over that can later become a standard PL example for Hydra users?

A good base would be to take one of the existing examples [here](https://github.com/PyTorchLightning/pytorch-lightning/tree/master/pl_examples/basic_examples).
Ideally we would have Hydra ports for all of these that shows the user how to do the same with Hydra.

if you want to do that, I am happy to answer questions and offer advice.

On a site note:

One thing that strikes me as odd in this PR is that we only worry about --gpus.
Why is it different than other command line argument the user is passing? shouldn't the ddp spawns get **everything** the user passed in the command line?
it doesn't matter if it's `--gpus=4` or `gpus=4` which PL seems to know about or something else like `db.user=root` or `optimizer.lr=0.01`. PL should pass it all when it spawns a new process for ddp.

Reply:
Yes, I would be able to help out with some of the integration here. I have created an example of using Hydra with the cpu template [here](https://github.com/anthonytec2/pytorch-lightning/blob/master/pl_examples/basic_examples/cpu_template.py). The general structure is, I have a main configuration file with some trainer settings and have a models directory. I expand the models parameters into the model and expand into trainer those specific settings.  Would it be better to use a structured config of the model and pass this directly into the model? Let me know should you think of any other improvements.

Reply:
Awesome, I commented on your commit.

Reply:
Thanks, I just fixed your comments and added a commit on supporting trainer arguments.